### DISCUSSION TITLE
Lazer using a different poll rate than the OS

### DISCUSSION THREAD

**Discussion by SevenChords** [ppy/osu#19918] [Q&A] [answered]
I noticed that my keyboard (supposedly 1000Hz poll rate according to the technical specifications provided by the manufacturer) runs at a poll rate of 120Hz in osu (hit error meter showing clearly that all lines are grouped at 120bpm/240bpm songs, tokaku made a video a while back that if you got a lower poll rate you will see those lines when playing songs that have their bpm set to a multiple of your poll rate). I tried some of those online scan tools, and got different results on different tests, but never the 120Hz osu lazer seems to use (got results of 250Hz, 333Hz and 1000Hz, not really reliable results i'd say, but i imagined them being at least closely similar to what lazer receives). i also tried a different keyboard which got very similar to the same results on the online scan rate tests, and no low-poll-rate lines in lazer, so i'm a bit confused as my main keyboard basically suggests it's the game doing something weird while another keyboard suggests it's not the game but rather the keyboard. I also tried different fps limit settings, didn't change anything when i cycled through all of them.

Any ideas as to what is going on and how i could get the game to recognise/use the 1000Hz poll rate from my keyboard would be great. If that is somehow of relevance, i am on macos, version 12.5, game version is 2022.821.0-lazer.

**Comment by peppy** [comment, marked as answer]
It is based on the game's update frame rate for now. Eventually it will be improved further to be more coupled to the input polling rate.

You shouldn't worry too much about this.

**Reply by SevenChords** [reply]
that does explain what i see when using my main keyboard, but not that i get a different poll rate with a different keyboard (used that other keyboard on a lot of different bpm to try and find out what poll rate i am getting with that one, no success)
edit: nvm, it does explain everything, my other keyboard is probably 125Hz, and with the game limiting to 120Hz and the kb still using 125Hz i ofc wont notice that within beatmaps

**Reply by SevenChords** [reply]
posting a new reply since i now got a new monitor with higher refresh rate, the update rate shown under the ctrl+f11 display shows as 1000Hz (fps limited to 4x refresh rate), yet the game is still limiting to 120Hz keyboard poll rate, so if it is supposed to be based on the update rate that part isn't working, as i got 1000Hz update rate shown but still getting the 120Hz poll rate (very clearly visible on 120bpm and 240bpm beatmaps in the hit error bar, all lines are in a few spots)

**Reply by peppy** [reply]
No idea sorry. You're going to have to provide video with the frame graphs visible showing what you mean to get anywhere (the video should compare the two different 120/240 polling or whatever. i have no idea what you're seeing and it's probably a red herring).

**Reply by SevenChords** [reply]
Here is the video you asked for: https://streamable.com/th8mmh
The beatmap played in the video is at 240bpm, other than the game limiting to 120Hz poll rate while running at 1000Hz update rate i don't have any idea what could be causing the timing groups seen on the hit error display (got the poll rate idea for those weird timings from this video: https://www.youtube.com/watch?v=heZVmr9fyng).
I have confirmed with the manufacturer's support that my keyboard indeed does have 1000Hz poll rate (using a GMMK1).
I attached the logs from the logs folder as well in case those help.
Will be on holiday for a week, if you require any additional logs/screenshots/screen recordings i will post them once i get back.

[database.log](https://github.com/ppy/osu/files/9439447/database.log)
[network.log](https://github.com/ppy/osu/files/9439448/network.log)
[performance.log](https://github.com/ppy/osu/files/9439449/performance.log)
[runtime.log](https://github.com/ppy/osu/files/9439450/runtime.log)

**Reply by peppy** [reply]
Try a different keyboard. It's probably the keyboard.

**Reply by SevenChords** [reply]
i only have a 125Hz keyboard lying around besides the one i am using (and ofc with 125Hz poll rate on the keyboard i can't see grouped timing lines while the game is running a 120Hz poll rate, which is still an assumption, got no way to confirm that fact to be 100%, it's just what i thought was happening based on the lines and tokaku's video), and one of these online scan rate testing sites did recognise a 1ms long key press from this keyboard (was just spamming a ton of keys at once to get more chances at a shorter press time), if there were software to properly test the poll rate of a keyboard available on macos (i couldn't find any myself, there may very well be such software) i'd be happy to try that, but i kind of don't want to spend money on another keyboard for this specific issue, and the support team from gmmk did confirm the keyboard itself is using a 1000Hz poll rate and the OS used shouldn't affect this

edit: just did the test on one of them online scan rate testing tools again, suggesting a 1000Hz scan rate, site and screenshot:
https://blog.seethis.link/scan-rate-estimator/
![Screenshot 2022-09-03 at 19 52 36](https://user-images.githubusercontent.com/31819497/188282584-4203d136-a009-4f77-a5bf-3ebdb48f1683.png)

**Reply by SevenChords** [reply]
final update:

while i have not been able to try out another keyboard, i have tried the same keyboard in two other games (the old osu client and quaver) and have been able to confirm what you are saying, it is the keyboard that is at fault here, and the manufacturer is falsely advertising/claiming 1000Hz poll rate in the technical specifications, which also shows that the scan rate estimator website is a bunch of bs too.
i will now be closing this discussion/marking it as answered, sorry for wasting your time on this topic and thanks for having the patience to deal with me on this.

**Reply by peppy** [reply]
Based on the way the site you are using works, it's possible that the keyboard is doing 1000hz down-up polling, but not 1000hz for new keypresses. If you read the methodology the site uses, it's about the time of a keydown event, not the actual rate of polling the keyboard for new keys.
